src/base/TextException.h

```diff
@@ -57,20 +58,18 @@ std::ostream &operator <<(std::ostream &, const TextException &);
 /// legacy convenience macro; it is not difficult to type Here() now
 #define TexcHere(msg) TextException((msg), Here())
 
-/// Like assert() but throws an exception instead of aborting the process
-/// and allows the caller to customize the exception message and location.
+/// Like Assure() but only logs the exception if level-3 debugging is enabled
+/// and runs even when NDEBUG macro is defined. Deprecated: Use Assure() for
+/// code logic checks and throw explicitly when input validation fails.
+#define Must(condition) \
+    Assure_(3, (condition), #condition, Here())
```

Comment:
> The removal is coming as a followup PR soon

I doubt we would be able to remove ~160 Must()s soon enough to justify code duplication here. Must() removal cannot be automated (because we cannot automatically distinguish code logic checks from input validation) and many existing input-validating Must()s deserve non-trivial code replacements.

On the other hand, if Must()s are indeed removed soon, I would be very pleased and will, of course, promptly volunteer the corresponding Assure() adjustments.

Either way, the outcome will be a good one!

Reply:
TL;DR - I do insist on this change. We will have to agree to disagree on our predictions of the future.

Long version;
Squid already has those Must() in existence. So this request is a matter of limiting the improvement/change to the set which can be converted to Assure(). To avoid the scope-creep of fixing code unrelated to Assure() itself.

(I see you here again falling for the trap of bundling multiple code improvements: Must() removal + protocol handling improvements. The total is great long-term, very tempting. But keeping the parts independent makes it easier to get the important parts of both done quicker.)

Reply:
> I do insist on this change. 

You made the right call, thank you. Please unblock this PR.

> Squid already has those Must() in existence. So this request is a matter of limiting the improvement/change to the set which can be converted to Assure(). To avoid the scope-creep of fixing code unrelated to Assure() itself.

I still see no good reason to duplicate this functionality, and I see no scope creep in this PR. I see in-scope refactoring of Must() _implementation_ to avoid significant (and costly) code/functionality duplication. This refactoring is the best way forward here IMO. Duplication is one of the primary sources of bugs and development problems in Squid. We should avoid making things worse. This PR avoids increasing duplication.

And just to clarify if somebody is reading this exchange too quickly: Leaving Must() "as is" in this PR will _not_ "limit improvements to the set of Must()s that will be converted to Assure()". Leaving Must() as is will leave _all_ Must()s "as is", including those Must()s that will be eventually converted to Assure(). Bugs notwithstanding, this PR makes the existing Squid code better, even though the new Assure() is still unused. That is a good thing! The alternative would make Squid code (slightly) worse (until we start using Assure()).

> I see you here again falling for the trap of bundling multiple code improvements

I do not see tightly-related improvements as a trap. In this specific case, I see them as the best way to make progress.

Reply:
> I do insist on this change.

> You made the right call, thank you. Please unblock this PR.

Ah, I misread. Sorry. I see no valid reason for you to insist that Must() and Assure() implementations must duplicate each other. Can you give one?

And given for how long this PR was mishandled, perhaps you can find a way to let it merge despite your reservations?

Reply:
> Ah, I misread. Sorry. I see no valid reason for you to insist that Must() and Assure() implementations must duplicate each other. Can you give one?
> 

I am **not** insisting that. I am requesting that the existing Must() definition be left **untouched** by this PR.

AFAICS, the new Assure() definition is different from the existing/legacy Must() definition. New debugs() labels, new message syntax, new internal implementation.

I have several reasons:

 1) the current (mis)users of Must() were written assuming / relying on its existing behaviour and output.
 2) other followup PRs will be cleaning/removing bad uses of Must().
 3) linking the two definitions is polluting/degrading the new Assure() API implementation.
 3) linking the two definitions is polluting/degrading the old Must() log outputs.
 4) we do not need to change it

Reply:
> > Alex: I see no valid reason for you to insist that Must() and Assure() implementations must duplicate each other.

> Amos: I am **not** insisting that. I am requesting that the existing Must() definition be left **untouched** by this PR.

... which will necessarily (poorly) duplicate Must() and Assure() implementations. On rare occasions, code duplication is the lesser of two evils, but I see no sufficient justification for implementing the same thing twice here:

* Must() implementation is check-condition-and-throw-exception code
* Assure() implementation is check-condition-and-throw-exception code

The only differences between the two correct implementations are the debug levels (0 vs. 3), and that the level-0 message needs a "ERROR: Squid BUG: " label. All the implementation code that is difficult to get right is actually identical.


> AFAICS, the new Assure() definition is different from the existing/legacy Must() definition. New debugs() labels, new message syntax, new internal implementation.

Yes, the message level and the message label/prefix naturally differ between Assure() and Must(), but that trivial difference can be (and is) expressed in one internal function parameter. We should not duplicate non-trivial internal code just because Assure() and Must() have different message levels (and because level-0 messages need a classification label).

> I have several reasons:

> 1. the current (mis)users of Must() were written assuming / relying on its existing behaviour and output.

This is not a valid reason because that old Must() behavior is preserved by this PR. 

Also, since branch commit 82817f6, the old level-0/1 output is also preserved. Here is a sample of the two failed Must(fail) checks that lead to level-0/1 messages, one in the official code and one in the PR branch code:


Before this PR:

```
2022/03/18 10:56:08.846| FATAL: check failed: false
    exception location: main.cc(1706) SquidMain
```

After PR commit 82817f6:

```
2022/03/23 17:37:15.791| FATAL: check failed: false
    exception location: main.cc(1661) SquidMain
```

Debugging (i.e. level-3) output has a different reporting location, of course, but is otherwise unchanged. Here is a sample:

Before this PR:

```
2022/03/18 10:56:08.846| 0,3| main.cc(1706) SquidMain: check failed: false
    exception location: main.cc(1706) SquidMain
```

After PR commit 82817f6:

```
2022/03/23 17:37:15.791| 0,3| Assure.cc(21) ReportAndThrow_: check failed: false
    exception location: main.cc(1661) SquidMain
```

Prior to branch commit 82817f6, the "check failed" prefix was not a part of the TextException text, but that accidental omission was my bug -- that text should be preserved, of course.


> 2. other followup PRs will be cleaning/removing bad uses of Must().

This is not a valid reason because this PR has no effect on cleaning/removing bad uses of Must(). When the very last Must() caller is gone, we will simplify Assure() implementation, but that change itself is simple, and I have already volunteered to do it.

> 3. linking the two definitions is polluting/degrading the new Assure() API implementation.

True, but having the level parameter and a simple extra condition (to label a level-0 message) is much better than poorly duplicating Must() and Assure() implementation, especially when that condition is not on a performance sensitive path and has a 5% binary size reduction side effect.

> 4. linking the two definitions is polluting/degrading the old Must() log outputs.

This is not a valid reason because this PR has no such effects on Must() log output.

> 5. we do not need to change it

This is not a valid reason because the need for that change has been established: Prevention of (poor) code duplication (and binary size reduction as a nice side effect of a better implementation).

In summary, we should not block a mishandled PR because it trades poor code duplication (and binary bloat) for an extra simple internal parameter/condition (outside performance sensitive paths).